### DISCUSSION TITLE
Raspberry pico support?

### DISCUSSION THREAD

**Discussion by cocasema** [modm-io/modm#752] [Ideas]
Wondering if there are any plans to support raspberry pico?

**Comment by salkinium** [comment]
Not from my side at the moment, although there was interest by @rleh I think?

**Reply by salkinium** [reply]
There's also no explicit support for dual-core devices in the build system. You can of course create two separate applications for each core, but inter-processor communications has to be done manually. I'm looking into adding support for dual-core H7, so that will help with other dual-core devices.

**Reply by rleh** [reply]
The Raspberry pico chip is certainly an interesting target; but I have neither concrete plans, nor time and motivation to implement support for the chip in the foreseeable future.

We are of course always happy about contributions to modm! 😉

**Comment by cocasema** [comment]
Not sure if it's a good idea to spam Discussions with separate questions related to RP2040 support, so decided to resume this old generic topic. Please let me know if you'd prefer to not to mix multiple topics into this one discussion.

**Reply by salkinium** [reply]
Go ahead, this topic is pretty generic. The discussion forum is also more wild west than the PRs or even issues, so…

**Comment by cocasema** [comment]
I'm trying to bring up [Adafruit Feather RP2040](https://learn.adafruit.com/adafruit-feather-rp2040-pico/overview), but no luck so far.

From the official SDK I see a few differences mentioned here (ignoring the default pins etc):

https://github.com/raspberrypi/pico-sdk/blob/2062372d203b372849d573f252cf7c6dc2800c0a/src/boards/include/boards/pico.h
vs
https://github.com/raspberrypi/pico-sdk/blob/2062372d203b372849d573f252cf7c6dc2800c0a/src/boards/include/boards/adafruit_feather_rp2040.h

To get started I modified a few rp2040 files (not adding new board variant until it works):

in modm repo

ext/modm-devices/devices/rp/rp2040.xml

```
      <memory name="flash" access="rx" start="0x10000000" size="8388608"/>  <<< 8MiB instead of 2
```

in my codebase in the generated files

updated project.xml

```
    <option name="modm:platform:core:boot2">generic_03h</option>
```

checked that FLASH size updated after lbuild build

modm/SConscript
```
env["CONFIG_DEVICE_MEMORY"] = [
    {'name': 'flash', 'access': 'rx', 'start': 268435456, 'size': 8388608}, <<<
    {'name': 'ram', 'access': 'rwx', 'start': 536870912, 'size': 262144},
    {'name': 'core1', 'access': 'rwx', 'start': 537133056, 'size': 4096},
    {'name': 'core0', 'access': 'rwx', 'start': 537137152, 'size': 4096},
]
```

modm/link/linkerscript.ld
```
MEMORY
{
	FLASH (rx) : ORIGIN = 0x10000000, LENGTH = 8388608 <<<<
```

modm/platform/clock/clocks.cpp

```
#ifndef PICO_XOSC_STARTUP_DELAY_MULTIPLIER
#define PICO_XOSC_STARTUP_DELAY_MULTIPLIER 64     <<<< changed from 1
#endif
```

modm/board/board.hpp

```
using LedGreen = GpioOutput13;    <<< instead of 25
```

... and not sure what/where to change to reflect PICO_FLASH_SPI_CLKDIV 2 --> 4

Still can't blink the LED. Could you please shed some light here..?

**Reply by rleh** [reply]
> 8MiB [flash] instead of 2

That is true, but to keep it simple I would suggest to ignore the larger flash at first. Not accessing the full flash does not hurt and we can solve this problem later. I'm afraid that the 2 MiB flash size might be implicitly assumed elsewhere and there might be problems to change the flash size in the device file.
(Also, I don't see why the size of the external flash chip should be hardcoded in the device file, but that's another issue...)

---

> `<option name="modm:platform:core:boot2">generic_03h</option>`

Pico-SDK does the same, but according to [Adafruits documentation](https://learn.adafruit.com/adafruit-feather-rp2040-pico/pinouts#microcontroller-and-flash-3084844) there is a QSPI flash on the board, which would be much faster than the generic _03h_ flash driver.
Could you visually identify the flash chip on your _Adafruit Feather RP2040_ board?

But assuming that the `generic_03h` driver works in general (this was probably never tested before in modm) this should not be a problem.

---

> `#define PICO_XOSC_STARTUP_DELAY_MULTIPLIER 64     <<<< changed from 1`

This define should go into [`device.hpp`](https://github.com/modm-io/modm/blob/develop/ext/rp/device.hpp.in#L19-L21) (somehow [via lbuild module config/options](https://github.com/modm-io/modm/blob/develop/ext/rp/module.lb#L33)), but your solution should work for now.

---

> `using LedGreen = GpioOutput13; // instead of 25`

👍🏽 

---

> ... and not sure what/where to change to reflect PICO_FLASH_SPI_CLKDIV 2 --> 4

The SPI CLKDIV cannot be changed easily since the [modm-ext/pico-sdk-partial](https://github.com/modm-ext/pico-sdk-partial) repo only contains prebuild bootloader blobs. (They are compiled [here](https://github.com/modm-ext/pico-sdk-partial/blob/main/update.py#L82).)
But for the `generic_03h` driver `PICO_FLASH_SPI_CLKDIV` [defaults to `4`](https://github.com/raspberrypi/pico-sdk/blob/master/src/rp2_common/boot_stage2/boot2_generic_03h.S#L33), so that should not be an issue.

In case you can't identify the flash chip on your board: Could you try all available `boot2` options (`[at25sf128a, generic_03h, is25lp080, w25q080, w25x10cl]`) and check if any of them works?

---

I don't have the Adafruit board laying around; maybe I'll order one tomorrow...

**Comment by cocasema** [comment]
Thanks a lot for the quick reply!

The flash chip seems to be [GD25Q64CSIG](https://www.mouser.com/ProductDetail/GigaDevice/GD25Q64CSIG)

![](https://ibin.co/6r2BG41pFRtM.jpg)

After removing all changes other than `#define PICO_XOSC_STARTUP_DELAY_MULTIPLIER 64` and `GpioOutput13` I got blinky working! :smile: 

And strangely it seems any of `.boot2` the options work fine... Although I found that changing project.xml and `lbuild build && scons uf2` wasn't enough to get the boot2.cpp file updated to the correct version, and I had to delete the `build` directory (I got spoiled by Bazel and keep forgetting about `cleaning` stage at all :stuck_out_tongue: )

**Reply by rleh** [reply]
Would you like to add a board support module for the Addafriut Feather RP2040 board to modm?
Would be 95% copy-paste of the existing [RP-Pico board module](https://github.com/modm-io/modm/tree/develop/src/modm/board/rp_pico) plus your changes (and hopefully in future with complete 8MiB flash usable)?

**Reply by cocasema** [reply]
I'd be happy to help. Btw, is there any way to inherit "properties" from the base/reference board? Or is it literally a copy-paste of the xml files? Looking at the pico-sdk/boards I see lots of variants with the same chip and very minor changes. Would be great to define rp2040 family of boards in some way...

**Reply by salkinium** [reply]
There are ways to inherit between boards, but copy-pasting is still recommended as the config will be displayed in the docs here: https://modm.io/reference/config/modm-rp-pico/#configuration

**Reply by cocasema** [reply]
got it :+1:

**Comment by cocasema** [comment]
![img](https://i.imgur.com/2iYM09T.jpeg)

**Comment by cocasema** [comment]
I want to add ADC support for rp2040. I have a POC with basic functionality working, but I'm stuck at finding and generating the needed info in the device tree. I can't figure out how to get get ADC support data for the pins from the [rp2040.svd](https://raw.githubusercontent.com/raspberrypi/pico-sdk/master/src/rp2040/hardware_regs/rp2040.svd) (Ctrl+F `GPIO26_CTRL`)
Pins [26..29] can be used as [A0..A3] inputs, but `FUNCSEL` is not helpful. When they are used as input, the AF is set to `null/31`, but that is also available as AF of all other pins...

I can get the ADC code generated and working if I manually update the generated rp2040.xml by adding
```
      <gpio port="bank0" pin="26" name="26">
+++    <signal driver="adc" name="in0" af="-1"/> 
        ....
```

So what would be the best way to fix this if the data is indeed missing? Should I update `tools/generator/rp_generator.py` to patch the svd file, like it's done for some STM32 devices?

**Reply by salkinium** [reply]
Yes, but I would just manually add the entries in Python, rather than patch the SVD file.

**Reply by salkinium** [reply]
Somewhere around here I would guess: https://github.com/modm-io/modm-devices/blob/develop/tools/generator/dfg/rp/rp_device_tree.py#L221
